### PR TITLE
Add an admin APIs to allow server admins to list users' pushers

### PR DESCRIPTION
**Pull request by dklimpel**
Add an admin API `GET /_synapse/admin/v1/users/<user_id>/pushers`
like https://matrix.org/docs/spec/client_server/latest#get-matrix-client-r0-pushers

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: synapse/rest/admin/users.py [hunk 1]
```diff
@@ -708,3 +719,29 @@ async def on_GET(self, request, user_id):
 
         ret = {"joined_rooms": list(room_ids), "total": len(room_ids)}
         return 200, ret
+
+
+class PushersRestServlet(RestServlet):
```
Comment (erikjohnston) [inline synapse/rest/admin/users.py:724]: Can you add a docstring to briefly explain what it is and the request/response please?

File: synapse/rest/admin/users.py [hunk 2]
```diff
@@ -708,3 +719,29 @@ async def on_GET(self, request, user_id):
...
+    PATTERNS = admin_patterns("/users/(?P<user_id>[^/]*)/pushers$")
+
+    def __init__(self, hs):
+        self.is_mine = hs.is_mine
+        self.store = hs.get_datastore()
+        self.auth = hs.get_auth()
+
+    async def on_GET(self, request, user_id):
```
Comment (erikjohnston) [inline synapse/rest/admin/users.py:732]: We've been really bad at asking for this, but can you add type hints here please? The idea is really that new code should be typed where possible

```suggestion
    async def on_GET(self, request: SynapseRequest, user_id: str) -> Tuple[int, JsonDict]:
```

Or something, you'll need to import those types.

File: synapse/rest/admin/users.py [hunk 3]
```diff
@@ -37,6 +37,17 @@
 
 logger = logging.getLogger(__name__)
 
+ALLOWED_KEYS = {
```
Comment (erikjohnston) [inline synapse/rest/admin/users.py:40]:
```suggestion
_GET_PUSHERS_ALLOWED_KEYS = {
```

File: synapse/rest/admin/users.py [hunk 4]
```diff
@@ -708,3 +719,29 @@ async def on_GET(self, request, user_id):
...
+        if not self.is_mine(UserID.from_string(user_id)):
+            raise SynapseError(400, "Can only lookup local users")
+
+        if not await self.store.get_user_by_id(user_id):
+            raise NotFoundError("User not found")
+
+        pushers = await self.store.get_pushers_by_user_id(user_id)
+
+        filtered_pushers = [
+            {k: v for k, v in p.items() if k in ALLOWED_KEYS} for p in pushers
```
Comment (erikjohnston) [inline synapse/rest/admin/users.py:744]:
```suggestion
            {k: v for k, v in p.items() if k in _GET_PUSHERS_ALLOWED_KEYS} for p in pushers
```

**Review by erikjohnston — CHANGES_REQUESTED** [review]
Looks good, just some nits.

**Comment by erikjohnston** [comment]
Thanks! Looks like there is a merge conflict?

**Comment by dklimpel** [comment]
> Thanks! Looks like there is a merge conflict?

I have solved the conflict. To many merges at the same day.  😄

**Review by erikjohnston — APPROVED** [review]

**Comment by erikjohnston** [comment]
> > Thanks! Looks like there is a merge conflict?
> 
> I have solved the conflict. To many merges at the same day.

There's been so many good PRs that I completely forgot I hadn't merged this one :D

**Pull request merged**